MinecraftClient/Mapping/Movement.cs

```diff
@@ -129,62 +129,89 @@ public static Queue<Location> Move2Steps(Location start, Location goal, ref doub
+
+            Location current = new Location(); // Location that is currently processed
+            Location closestGoal = new Location(); // Closest Location to the goal. Used for approaching if goal can not be reached or was not found.
+            HashSet<Location> ClosedSet = new HashSet<Location>(); // The set of locations already evaluated.
+            HashSet<Location> OpenSet = new HashSet<Location>(new[] { start });  // The set of tentative nodes to be evaluated, initially containing the start node
+            Dictionary<Location, Location> Came_From = new Dictionary<Location, Location>(); // The map of navigated nodes.
+
+            Dictionary<Location, int> g_score = new Dictionary<Location, int>(); //:= map with default value of Infinity
+            g_score[start] = 0; // Cost from start along best known path.
+            // Estimated total cost from start to goal through y.
+            Dictionary<Location, int> f_score = new Dictionary<Location, int>(); //:= map with default value of Infinity
+            f_score[start] = (int)start.DistanceSquared(goal); //heuristic_cost_estimate(start, goal)
+
+            while (OpenSet.Count > 0)
             {
-                HashSet<Location> ClosedSet = new HashSet<Location>(); // The set of locations already evaluated.
-                HashSet<Location> OpenSet = new HashSet<Location>(new[] { start });  // The set of tentative nodes to be evaluated, initially containing the start node
-                Dictionary<Location, Location> Came_From = new Dictionary<Location, Location>(); // The map of navigated nodes.
+                current = //the node in OpenSet having the lowest f_score[] value
+                    OpenSet.Select(location => f_score.ContainsKey(location)
+                    ? new KeyValuePair<Location, int>(location, f_score[location])
+                    : new KeyValuePair<Location, int>(location, int.MaxValue))
+                    .OrderBy(pair => pair.Value).First().Key;
+
+                // Only assert a value if it is of actual use later
+                if (maxOffset > 0 && ClosedSet.Count > 0)
+                    closestGoal = ClosedSet.OrderBy(checkedLocation => checkedLocation.DistanceSquared(goal)).First();
 
-                Dictionary<Location, int> g_score = new Dictionary<Location, int>(); //:= map with default value of Infinity
-                g_score[start] = 0; // Cost from start along best known path.
-                // Estimated total cost from start to goal through y.
-                Dictionary<Location, int> f_score = new Dictionary<Location, int>(); //:= map with default value of Infinity
-                f_score[start] = (int)start.DistanceSquared(goal); //heuristic_cost_estimate(start, goal)
+                // Stop when goal is reached or we are close enough
+                if (current == goal || (current.DistanceSquared(goal) <= minOffset && minOffset > 0))
```

Comment:
Skip calculating `DistanceSquared()` if not needed:
```suggestion
                if (current == goal || ( minOffset > 0 && current.DistanceSquared(goal) <= minOffset))
```